### ISSUE TITLE
Generation of yrep for predictive checks leaks information

### ISSUE THREAD

**Issue by ErikRingen** [ScottClaessens/coevolve#72]
While doing a new round of SBC I realized that the current formulation of the predictive checks via yrep will be overly optimistic because it uses information from other trait values to predict observations at the tips, when in fact those values should be generated jointly/simultaneously. The only change that needs to occur is in the generated quantities block:

```stan
generated quantities{
  vector[N_obs*J] log_lik; // log-likelihood
  array[N_tree,N_obs,J] real yrep; // predictive checks
  matrix[J,J] cor_R; // correlated drift
  cor_R = multiply_lower_tri_self_transpose(L_R);
  {
    matrix[N_obs,J] log_lik_temp = rep_matrix(0.0, N_obs, J);
    for (i in 1:N_obs) {
      array[N_tree,N_obs,J] real lp = rep_array(0.0, N_tree, N_obs, J);
      for (t in 1:N_tree) {
        vector[J] mu_cond;
        vector[J] sigma_cond;
        vector[J] residuals;
        matrix[J,J] cov_inv = inverse_spd(VCV_tips[t, tip_id[i]]);
        vector[J] terminal_drift_rep;
        
        for (j in 1:J) terminal_drift_rep[j] = normal_rng(0, 1);
        terminal_drift_rep = cholesky_decompose(VCV_tips[t, tip_id[i]]) * terminal_drift_rep;
        
        residuals[1] = y[i][1] - eta[t,tip_id[i]][1];
        residuals[2] = terminal_drift[t, tip_id[i]][2];
        mu_cond = residuals - (cov_inv * residuals) ./ diagonal(cov_inv);
        sigma_cond = sqrt(1 / diagonal(cov_inv));
        if (miss[i,1] == 0) lp[t,i,1] = normal_lpdf(residuals[1] | mu_cond[1], sigma_cond[1]);
        yrep[t,i,1] = eta[t,tip_id[i]][1] + terminal_drift_rep[1];
        if (miss[i,2] == 0) lp[t,i,2] = bernoulli_logit_lpmf(to_int(y[i,2]) | eta[t,tip_id[i]][2] + residuals[2]);
        yrep[t,i,2] = bernoulli_logit_rng(eta[t,tip_id[i]][2] + terminal_drift_rep[2]);
      }
    for (j in 1:J) log_lik_temp[i,j] += log_sum_exp(lp[,i,j]);
    }
  log_lik = to_vector(log_lik_temp);
  }
}

```

**Comment by ScottClaessens** [comment]
Thanks for spotting this. I will take a look soon.

**Comment by ErikRingen** [comment]
Thanks! For context I realized that this was a similar problem to, for example, when one does a Poisson with an observation-level random effect (OLRE) to soak up overdispersion. When doing predictive checks in such a case one needs to draw new levels of the OLRE, or else the checks will be way over-optimistic.

**Comment by ScottClaessens** [comment]
To be clear about the changes, none of this is necessary if there are no normally distributed variables, right?

**Comment by ErikRingen** [comment]
> @ErikRingen To be clear about these changes, none of this is necessary if there are no normally distributed variables, right?

I am pretty sure for all variables, actually. The same logic as the Gaussian, just with the additional layer of the link function.

**Comment by ErikRingen** [comment]
Yeah I guess it implies additional re-factoring of the code. But I am (fairly) certain that what I am suggesting is the right way to do this. For example if it was a typical multi-response regression model, where lets say all the responses are bernoulli, then you would want to re-sample the correlated OLREs before generating yrep. Because from the perspective of the generative model, all those responses are generated simultaneously.

**Comment by ScottClaessens** [comment]
Okay, I think I understand. Basically, we want to ensure that information from _all_ traits is used in the predictions, and the worry is that the `eta` values alone don't provide that cross-trait information.

Is it almost like replaying the terminal branches of the tree again, but allowing the drift to be different this time?

**Comment by ScottClaessens** [comment]
So, we had made a change before to remove some unneeded complexity for models without any Gaussian variables. However, I think that making this change to the generated quantities block requires us to go back to the previous implementation, where the terminal drift is added in the model block for all variable types (Gaussian and non-Gaussian). This is because the `VCV_tips` and `residuals` parameters are required in the generated quantities block.

I've opened up a branch for this issue and pushed some initial changes to `coev_make_stancode()`. @ErikRingen when you get a sec, could you take a look at the diff for this commit and maybe try generating some Stan code to see if this deals with the issue in the correct way? It would probably be good to look at various mixtures of Bernoulli and Gaussian variables.

I haven't fitted any models with this code yet, as I'm on my laptop today. Once we're happy with the changes to `coev_make_stancode()`, I will implement unit tests and finalise.

**Comment by ErikRingen** [comment]
Just checked it out, all looks good to me.

**Comment by ScottClaessens** [comment]
Thanks! Were you able to fit any models with this Stan code?

**Comment by ErikRingen** [comment]
Yes, was able to successfully fit and do predictive checks for a model with one gaussian one ordered.

**Comment by ScottClaessens** [comment]
Thank you! I'll work on the remaining changes soon. Trying to prioritise the updated vignette for Thursday's workshop. I definitely want to show off your time series plots and our multiPhylo implementation.

**Comment by ErikRingen** [comment]
One note after looking at some of the stancode again:

This part is not needed in the generated quantities block when log_lik = F:

```
        matrix[J,J] cov_inv = inverse_spd(VCV_tips[t, tip_id[i]]);
        mu_cond = residuals - (cov_inv * residuals) ./ diagonal(cov_inv);
        sigma_cond = sqrt(1 / diagonal(cov_inv));
```

**Comment by ScottClaessens** [comment]
Thanks for catching that! I've pushed another change to the branch to deal with this.

**Comment by ScottClaessens** [comment]
This issue should be fixed in f7eb6ec.

**Comment by ScottClaessens** [comment]
Just reopening this issue.

A colleague got in touch with me today to say that a model that was working before the change in f7eb6ec is now no longer converging properly. After doing some digging, I think it has something to do with the changes that we introduced on this branch, where we include `residuals` and `VCV_tips` even for non-Gaussian variables. This addition seems to have made my own models more inefficient as well (e.g., the authority model).

I wonder if there is another way to approach this predictive check problem that avoids adding more complexity than necessary when there are no Gaussian variables. One approach could be to go back to the version that we had before (i.e., final drift segment is added in the transformed parameters block rather than the model block for non-Gaussian variables), but we use the _parent nodes_ to jointly simulate new observations at the tips? Basically repeating this section of the transformed parameters block:

```stan
to_vector(A_delta * eta[t, parent[t, i]] + ((A \ add_diag(A_delta, -1)) * b) + drift_seg)
```

in the generated quantities block. This reduces complexity in the actual fitted model and deals with the leaking of information for the predictions. What do you think @ErikRingen?

**Comment by ErikRingen** [comment]
Hey, I think that the only substantive difference for non-Gaussian traits would be the use of non-centered parameterization. So when there are no Gaussian traits we can do:

terminal_drift ~ std_normal()
residuals = terminal_drift * cholesky_decompose(VCV_tips)

This could be done in either the transformed parameters block or the model block, I think the former is more efficient.

**Comment by ScottClaessens** [comment]
This sounds like a good suggestion. This would be only for models where there are no Gaussian traits, right? I can trying implementing this soon.

**Comment by ErikRingen** [comment]
Yep, only when there are non-Gaussian traits (because of the way the likelihood needs to be factorized).

**Comment by ScottClaessens** [comment]
@ErikRingen for a Bernoulli-Ordinal model, does this updated model block look right to you? I've implemented the non-centred parameterisation, as you suggested. I fitted the model to simulated data and it seems to fit smoothly.

```stan
model{
  b ~ std_normal();
  for (t in 1:N_tree) {
    eta_anc[t] ~ std_normal();
    for (i in 1:(N_seg - 1)) z_drift[t, i] ~ std_normal();
    for (i in 1:N_obs) to_vector(terminal_drift[t][i,]) ~ std_normal();
  }
  A_offdiag ~ std_normal();
  A_diag ~ std_normal();
  L_R ~ lkj_corr_cholesky(4);
  Q_sigma ~ std_normal();
  c2 ~ normal(0, 2);
  if (!prior_only) {
    for (i in 1:N_obs) {
      vector[N_tree] lp = rep_vector(0.0, N_tree);
      for (t in 1:N_tree) {
        vector[J] residuals;
        residuals = cholesky_decompose(VCV_tips[t, tip_id[i]]) * to_vector(terminal_drift[t][i,]);
        if (miss[i,1] == 0) lp[t] += bernoulli_logit_lpmf(to_int(y[i,1]) | eta[t,tip_id[i]][1] + residuals[1]);
        if (miss[i,2] == 0) lp[t] += ordered_logistic_lpmf(to_int(y[i,2]) | eta[t,tip_id[i]][2] + residuals[2], c2);
      }
      target += log_sum_exp(lp);
    }
  }
}
```

**Comment by ErikRingen** [comment]
Yep, this is what I had in mind.

**Comment by ScottClaessens** [comment]
Implemented the non-centered parameterisation in d305614. Hopefully this fixes the issue!
